### DISCUSSION TITLE
Batch tensor not being put on cuda

### DISCUSSION THREAD

**Discussion by davidireland-iso** [pyg-team/pytorch_geometric#3299] [Q&A] [answered]
I am using top_k pooling with GPU but having an issue where the batch tensor which tells which graph the nodes belong to, from the data loader, is not being put on the GPU. 

The code looks like 
```
    def forward(self, data, return_node_embedding=False, return_both=False):
        x, edge_index, batch = data.x, data.edge_index, data.batch  # here data.batch is not on the gpu like the x and edge index tensors

        x = torch.relu(self.gcn1(x, edge_index))
        if return_node_embedding and not return_both:
            return x
        elif return_both:
            x_ = copy.deepcopy(x.detach())
        x, edge_index, _, batch, _, _ = self.first_pool_layer(x, edge_index, batch=batch) 
```
 The data comes from a DataLoader using `train_loader = DataLoader(train_data, batch_size=128, shuffle=True)`. 

Has anyone had this issue? 

e: this has happened since updating to the latest version of pyg.

**Comment by rusty1s** [comment, marked as answer]
This should be already fixed in master, see https://github.com/pyg-team/pytorch_geometric/pull/3214.

**Reply by davidireland-iso** [reply]
Thanks -- do you know how I can get this version on my machine? I only used `conda` to update to the latest version of pyg yesterday and I'm having this issue.

**Reply by rusty1s** [reply]
You can install from master via `pip`: `pip install git+https://github.com/pyg-team/pytorch_geometric.git`

If you don't wanna upgrade, simple calling `batch = batch.to(device)` should do the trick as well.

**Reply by davidireland-iso** [reply]
thanks for the help!

I got my script to work, but as an aside I now have another question. has anything changed in the new version that would change the way `top_k_pool` or `SAGEConv` to act differently? I am getting quite different embeddings running the same script on the new version than I did on the previous.

**Reply by rusty1s** [reply]
What do you mean by getting different embeddings? Is there a drop in performance? AFAIK, there aren't any changes in `top_k_pool` or `SAGEConv` which shall explain your observations. Do you have a small script to reproduce?

**Reply by davidireland-iso** [reply]
sorry, I should have given more description of what I am doing. 
I essentially 4 classes for my dataset of graph and I am obtaining embeddings using a contrastive loss. Before I updated to the latest version of `pyg` I was getting a nice embedding space that was essentially separated into clusters. When I updated today and ran the same script, I was getting a weird shape of embedding where they were all essentially on a line (embedding space is 2-dimensional). However, I have since ran it after using your `pip` update from the previous comment and it is now back to giving a 'nice' embedding space again, so I'm not sure what was happening initially. 

On that note however, after I used the above `pip` command I was still getting the error that the batch tensor was not put onto the GPU.

**Reply by rusty1s** [reply]
Thanks for the clarification. Might have been a problem with some unlucky weight initialization?
Furthermore, can you give me a small example to reproduce the device issues?

**Reply by davidireland-iso** [reply]
```
from torch_geometric.loader import DataLoader
import torch
import torch.nn as nn
from torch_geometric.nn import global_mean_pool, global_max_pool, SAGEConv, TopKPooling
import numpy as np
import random
import copy

random.seed(1)
torch.manual_seed(1)
np.random.seed(1)

class KPooling(nn.Module):

    def __init__(self, ratio, input_size, hidden_size, output_size):

        super(KPooling, self).__init__()

        self.gcn1 = SAGEConv(input_size, hidden_size)
        self.first_pool_layer = TopKPooling(hidden_size, ratio)
        self.gcn2 = SAGEConv(hidden_size, hidden_size)
        self.second_pool_layer = TopKPooling(hidden_size, ratio)
        self.gcn3 = SAGEConv(hidden_size, hidden_size)
        self.third_pool_layer = TopKPooling(hidden_size, ratio)
        self.output_layer = nn.Linear(hidden_size * 2, output_size)

    def forward(self, data, return_node_embedding=False, return_both=False):
        x, edge_index, batch = data.x, data.edge_index, data.batch

        x = torch.relu(self.gcn1(x, edge_index))
        if return_node_embedding and not return_both:
            return x
        elif return_both:
            x_ = copy.deepcopy(x.detach())
        x, edge_index, _, batch, _, _ = self.first_pool_layer(x, edge_index, batch=batch) # this is the line where it breaks
        summary_1_mean = global_mean_pool(x, batch)
        summary_1_max = global_max_pool(x, batch)
        summary_1 = torch.cat((summary_1_mean, summary_1_max), dim=1)

        x = torch.relu(self.gcn2(x, edge_index))
        x, edge_index, _, batch, _, _ = self.second_pool_layer(x, edge_index, batch=batch)
        summary_2_mean = global_mean_pool(x, batch)
        summary_2_max = global_max_pool(x, batch)
        summary_2 = torch.cat((summary_2_mean, summary_2_max), dim=1)

        x = self.gcn3(x, edge_index)
        x, edge_index, _, batch, _, _ = self.second_pool_layer(x, edge_index, batch=batch)
        summary_3_mean = global_mean_pool(x, batch)
        summary_3_max = global_max_pool(x, batch)
        summary_3 = torch.cat((summary_3_mean, summary_3_max), dim=1)

        summary = summary_1 + summary_2 + summary_3

        summary = self.output_layer(summary)

        if not return_both:
            return summary
        elif return_both:
            return summary, x_

train_loader = DataLoader(train_data, batch_size=128, shuffle=True)
encoder = KPooling(ratio, 2, embedding_size, output_size).to(device)
optimiser = torch.optim.Adam(encoder.parameters())
for count, batch in enumerate(train_loader):
    optimiser.zero_grad()
    inputs = encoder.forward(batch)
```

This should reproduce the device error, I've commented on the first line it breaks in the `forward` of the network.

**Reply by rusty1s** [reply]
I think it is not possible to reproduce without the `train_data` logic. Is the `train_data` already sitting on GPU?

**Reply by davidireland-iso** [reply]
Yes, train data is already on the GPU. The usage is this:

```
data = [d.to(device) for d in data]
n = int(len(data) * train_ratio)
train_data = data[:n]
```

where the original data object is a list of graphs that I have saved in storage, where each graph is `torch_geometric.data.Data` object.

**Reply by rusty1s** [reply]
Are you sure you are running on PyG master? This works without problems for me (with the fix from #3214):
```python
from torch_geometric.loader import DataLoader
from torch_geometric.datasets import TUDataset

dataset = TUDataset('/tmp/TU', 'MUTAG')
dataset = [data.cuda() for data in dataset]

loader = DataLoader(dataset, batch_size=32)

for data in loader:
    print(data.x.device)
    print(data.batch.device)
```

**Reply by davidireland-iso** [reply]
as far as I know I was: I used the `pip` command that you shared in my conda environment so it should've been the master. 

also, in case you are interested, I was still having some problems with the embeddings I mentioned in an earlier comment but I narrowed the problem down to being on PyTorch version 1.9 (as opposed to 1.8) which I also updated at the same time as the new PyG version -- not sure why this caused the problem but at least I found it.

**Reply by rusty1s** [reply]
Can you still confirm that the above code snippet from me works as expected for you?

**Reply by davidireland-iso** [reply]
I can confirm it _doesn't_ work as intended. My output was:
```
cuda:0
cpu
cuda:0
cpu
cuda:0
cpu
cuda:0
cpu
cuda:0
cpu
cuda:0
cpu
```

**Reply by rusty1s** [reply]
I think your old `conda` installation of PyG is still the active one. You should try to uninstall it first (`conda uninstall pyg`), and re-install from `pip` again.

**Reply by davidireland-iso** [reply]
Ah yes, this fixed it! Thanks so much for the help :-)
